src/Storages/StorageMySQL.cpp

```diff
@@ -234,68 +234,92 @@ SinkToStoragePtr StorageMySQL::write(const ASTPtr & /*query*/, const StorageMeta
-        const String & remote_table = engine_args[2]->as<ASTLiteral &>().value.safeGet<String>();
-        const String & username = engine_args[3]->as<ASTLiteral &>().value.safeGet<String>();
-        const String & password = engine_args[4]->as<ASTLiteral &>().value.safeGet<String>();
-        size_t max_addresses = args.getContext()->getSettingsRef().glob_expansion_max_elements;
-
-        /// TODO: move some arguments from the arguments to the SETTINGS.
-        MySQLSettings mysql_settings;
+            engine_arg = evaluateConstantExpressionOrIdentifierAsLiteral(engine_arg, context_);
+
+        const auto & host_port = engine_args[0]->as<ASTLiteral &>().value.safeGet<String>();
+        size_t max_addresses = context_->getSettingsRef().glob_expansion_max_elements;
+
+        configuration.addresses = parseRemoteDescriptionForExternalDatabase(host_port, max_addresses, 3306);
+        configuration.database = engine_args[1]->as<ASTLiteral &>().value.safeGet<String>();
+        configuration.table = engine_args[2]->as<ASTLiteral &>().value.safeGet<String>();
+        configuration.username = engine_args[3]->as<ASTLiteral &>().value.safeGet<String>();
+        configuration.password = engine_args[4]->as<ASTLiteral &>().value.safeGet<String>();
+
+        if (engine_args.size() >= 6)
+            configuration.replace_query = engine_args[5]->as<ASTLiteral &>().value.safeGet<UInt64>();
+        if (engine_args.size() == 7)
+            configuration.on_duplicate_clause = engine_args[6]->as<ASTLiteral &>().value.safeGet<String>();
+    }
+
+    if (configuration.replace_query && !configuration.on_duplicate_clause.empty())
+        throw Exception(ErrorCodes::BAD_ARGUMENTS,
+                        "Only one of 'replace_query' and 'on_duplicate_clause' can be specified, or none of them");
+
+    return configuration;
+}
+
+
+void registerStorageMySQL(StorageFactory & factory)
+{
+    factory.registerStorage("MySQL", [](const StorageFactory::Arguments & args)
+    {
+        auto configuration = StorageMySQL::getConfiguration(args.engine_args, args.getLocalContext());
+
+        MySQLSettings mysql_settings; /// TODO: move some arguments from the arguments to the SETTINGS.
         if (args.storage_def->settings)
-        {
             mysql_settings.loadFromQuery(*args.storage_def);
-        }
 
         if (!mysql_settings.connection_pool_size)
             throw Exception("connection_pool_size cannot be zero.", ErrorCodes::BAD_ARGUMENTS);
 
-        auto addresses = parseRemoteDescriptionForExternalDatabase(host_port, max_addresses, 3306);
+        std::cerr << "user: " << configuration.username << ", pass: " << configuration.password << std::endl;
         mysqlxx::PoolWithFailover pool(
-            remote_database,
-            addresses,
-            username,
-            password,
+            configuration.database, configuration.addresses,
+            configuration.username, configuration.password,
```

Comment:
So in a nut shell this patch just hides some credentials.

But maybe predefined connections should also share pool (to add ability to limit total number of connections to the specific external connection). @kssenii what do you think?

Reply:
Sounds like a good idea. But now pools are only for mysql and postgresql and for mysql as I remember there already a singleton pool factory. 

But Maksim was going to make a common interface for all pools for external engines connections (that big PR which was made a long time ago and is still wip) and when it is finished (I really hope it eventually will be) -- it will be very convenient and nice looking to implement what you suggest.

Reply:
>and for mysql as I remember there already a singleton pool factory.

AFAICS it is used only for dictionaries

>But Maksim was going to make a common interface for all pools for external engines connections

Great! Can you give me a link? (I don't see anything relevant here https://github.com/ClickHouse/ClickHouse/pulls/kitaisreal)

Reply:
Thought it is https://github.com/ClickHouse/ClickHouse/pull/23060 but now looked at it and he did not move the interface out of only mysql related. I just feltl like he once mentioned that it would be applicable to other, looks like it is not the case now.